pkg/queue/breaker.go

```diff
@@ -144,6 +166,17 @@ type semaphore struct {
 	mux      sync.RWMutex
 }
 
+// acquireNonBlocking receives the token from the semaphore if there's one
+// otherwise an error is returned.
+func (s *semaphore) acquireNonBlocking(ctx context.Context) error {
```

Comment:
In semaphores, this is usually called `tryAcquire` and it returns a `bool` instead. I think that'd make some nicer semantics, also above? Returning an error when we can't reserve a slot feels a little harsh to me.